File: apps/hyperliquid-recorder/src/hyperliquid_recorder/health.py
from __future__ import annotations

import json
import threading
import time
from http import HTTPStatus
from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from hyperliquid_recorder.metrics import RecorderMetrics


class HealthState:
    def __init__(self, expected_coins: tuple[str, ...], stale_seconds: int) -> None:
        self._expected_coins = expected_coins
        self._stale_seconds = stale_seconds
        self._market_last_message: dict[str, float] = {}
        self._clickhouse_ok = False
        self._lock = threading.Lock()

    def set_market_seen(self, coin: str) -> None:
        with self._lock:
            self._market_last_message[coin] = time.time()

    def set_clickhouse_ok(self, healthy: bool) -> None:
        with self._lock:
            self._clickhouse_ok = healthy

    def is_ready(self) -> bool:
        now = time.time()
        with self._lock:
            if not self._clickhouse_ok:
                return False
            for coin in self._expected_coins:
                seen = self._market_last_message.get(coin)
                if seen is None or now - seen > self._stale_seconds:
                    return False
            return True

    def to_dict(self) -> dict[str, object]:
        now = time.time()
        with self._lock:
            stale = []
            for coin in self._expected_coins:
                seen = self._market_last_message.get(coin)
                if seen is None or now - seen > self._stale_seconds:
                    stale.append(coin)
            ready = self._clickhouse_ok and not stale
            return {
                "ready": ready,
                "clickhouse_ok": self._clickhouse_ok,
                "stale_markets": stale,
            }


def start_http_servers(
    health_port: int,
    metrics_port: int,
    metrics: RecorderMetrics,
    state: HealthState,
) -> None:
    class HealthHandler(BaseHTTPRequestHandler):
        def do_GET(self) -> None:
            if self.path == "/live":
                payload = b'{"status":"live"}'
                self.send_response(HTTPStatus.OK)
                self.send_header("Content-Type", "application/json")
                self.send_header("Content-Length", str(len(payload)))
                self.end_headers()
                self.wfile.write(payload)
                return

            if self.path == "/ready":
                ready = state.is_ready()
                body = json.dumps(state.to_dict()).encode("utf-8")
                self.send_response(HTTPStatus.OK if ready else HTTPStatus.SERVICE_UNAVAILABLE)
                self.send_header("Content-Type", "application/json")
                self.send_header("Content-Length", str(len(body)))
                self.end_headers()
                self.wfile.write(body)
                return

            self.send_response(HTTPStatus.NOT_FOUND)
            self.end_headers()

        def log_message(self, _format: str, *_args: object) -> None:
            return None

    class MetricsHandler(BaseHTTPRequestHandler):
        def do_GET(self) -> None:
            if self.path != "/metrics":
                self.send_response(HTTPStatus.NOT_FOUND)
                self.end_headers()
                return

            payload = metrics.to_prometheus_payload()
            self.send_response(HTTPStatus.OK)
            self.send_header("Content-Type", "text/plain; version=0.0.4")
            self.send_header("Content-Length", str(len(payload)))
            self.end_headers()
            self.wfile.write(payload)

        def log_message(self, _format: str, *_args: object) -> None:
            return None

    health_server = ThreadingHTTPServer(("0.0.0.0", health_port), HealthHandler)
    metrics_server = ThreadingHTTPServer(("0.0.0.0", metrics_port), MetricsHandler)

    threading.Thread(target=health_server.serve_forever, daemon=True).start()
    threading.Thread(target=metrics_server.serve_forever, daemon=True).start()

File: apps/hyperliquid-recorder/src/health.rs
use std::{
    collections::HashMap,
    sync::{Arc, Mutex},
    time::{Duration, Instant, SystemTime, UNIX_EPOCH},
};

use anyhow::Result;
use axum::{extract::State, http::StatusCode, response::IntoResponse, routing::get, Json, Router};
use serde::Serialize;
use tokio::task::JoinHandle;

use crate::metrics::RecorderMetrics;

#[derive(Clone)]
pub struct HealthState {
    expected_coins: Vec<String>,
    stale_seconds: u64,
    funding_poll_seconds: u64,
    funding_stale_seconds: u64,
    started_at: Instant,
    inner: Arc<Mutex<HealthInner>>,
}

#[derive(Default)]
struct HealthInner {
    market_last_message: HashMap<String, f64>,
    funding_last_event_ms: HashMap<String, u64>,
    clickhouse_ok: bool,
}

#[derive(Serialize)]
struct ReadyResponse {
    ready: bool,
    clickhouse_ok: bool,
    stale_markets: Vec<String>,
    stale_funding_markets: Vec<String>,
}

impl HealthState {
    pub fn new(
        expected_coins: Vec<String>,
        stale_seconds: u64,
        funding_poll_seconds: u64,
        funding_stale_seconds: u64,
    ) -> Self {
        Self {
            expected_coins,
            stale_seconds,
            funding_poll_seconds,
            funding_stale_seconds,
            started_at: Instant::now(),
            inner: Arc::new(Mutex::new(HealthInner::default())),
        }
    }

    pub fn set_market_seen(&self, coin: &str) {
        let mut inner = self.inner.lock().expect("health mutex poisoned");
        inner
            .market_last_message
            .insert(coin.to_string(), unix_seconds_now());
    }

    pub fn set_funding_event_seen(&self, coin: &str, funding_time_ms: u64) {
        let mut inner = self.inner.lock().expect("health mutex poisoned");
        inner
            .funding_last_event_ms
            .insert(coin.to_string(), funding_time_ms);
    }

    pub fn set_clickhouse_ok(&self, healthy: bool) {
        let mut inner = self.inner.lock().expect("health mutex poisoned");
        inner.clickhouse_ok = healthy;
    }

    pub fn is_ready(&self) -> bool {
        self.to_ready_response().ready
    }

    fn to_ready_response(&self) -> ReadyResponse {
        let inner = self.inner.lock().expect("health mutex poisoned");
        let now = unix_seconds_now();
        let elapsed_secs = self.started_at.elapsed().as_secs();
        let grace_secs = self.funding_poll_seconds.saturating_mul(2);

        let stale_markets = self
            .expected_coins
            .iter()
            .filter(|coin| {
                inner
                    .market_last_message
                    .get(*coin)
                    .map(|seen| now - *seen > self.stale_seconds as f64)
                    .unwrap_or(true)
            })
            .cloned()
            .collect::<Vec<_>>();

        let now_ms = (now * 1000.0) as u64;
        let stale_funding_markets = self
            .expected_coins
            .iter()
            .filter(|coin| match inner.funding_last_event_ms.get(*coin) {
                Some(event_ms) => {
                    now_ms.saturating_sub(*event_ms) / 1000 > self.funding_stale_seconds
                }
                None => elapsed_secs >= grace_secs,
            })
            .cloned()
            .collect::<Vec<_>>();

        ReadyResponse {
            ready: inner.clickhouse_ok
                && stale_markets.is_empty()
                && stale_funding_markets.is_empty(),
            clickhouse_ok: inner.clickhouse_ok,
            stale_markets,
            stale_funding_markets,
        }
    }
}

#[derive(Clone)]
struct HealthAppState {
    state: HealthState,
}

#[derive(Clone)]
struct MetricsAppState {
    metrics: Arc<RecorderMetrics>,
}

pub fn start_http_servers(
    health_port: u16,
    metrics_port: u16,
    metrics: Arc<RecorderMetrics>,
    state: HealthState,
) -> (JoinHandle<()>, JoinHandle<()>) {
    let health_handle = tokio::spawn(async move {
        if let Err(err) = run_health_server(health_port, state).await {
            tracing::error!("health server failed: {err:?}");
        }
    });

    let metrics_handle = tokio::spawn(async move {
        if let Err(err) = run_metrics_server(metrics_port, metrics).await {
            tracing::error!("metrics server failed: {err:?}");
        }
    });

    (health_handle, metrics_handle)
}

async fn run_health_server(port: u16, state: HealthState) -> Result<()> {
    let app = Router::new()
        .route("/live", get(live_handler))
        .route("/ready", get(ready_handler))
        .with_state(HealthAppState { state });
    axum::Server::bind(&format!("0.0.0.0:{port}").parse()?)
        .serve(app.into_make_service())
        .await?;
    Ok(())
}

async fn run_metrics_server(port: u16, metrics: Arc<RecorderMetrics>) -> Result<()> {
    let app = Router::new()
        .route("/metrics", get(metrics_handler))
        .with_state(MetricsAppState { metrics });
    axum::Server::bind(&format!("0.0.0.0:{port}").parse()?)
        .serve(app.into_make_service())
        .await?;
    Ok(())
}

async fn live_handler() -> impl IntoResponse {
    (
        StatusCode::OK,
        Json(serde_json::json!({ "status": "live" })),
    )
}

async fn ready_handler(State(app): State<HealthAppState>) -> impl IntoResponse {
    let response = app.state.to_ready_response();
    let status = if response.ready {
        StatusCode::OK
    } else {
        StatusCode::SERVICE_UNAVAILABLE
    };
    (status, Json(response))
}

async fn metrics_handler(State(app): State<MetricsAppState>) -> impl IntoResponse {
    match app.metrics.to_prometheus_payload() {
        Ok(payload) => (
            StatusCode::OK,
            [("content-type", "text/plain; version=0.0.4")],
            payload,
        )
            .into_response(),
        Err(err) => (
            StatusCode::INTERNAL_SERVER_ERROR,
            [("content-type", "text/plain; charset=utf-8")],
            format!("failed to encode metrics: {err}").into_bytes(),
        )
            .into_response(),
    }
}

fn unix_seconds_now() -> f64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap_or(Duration::ZERO)
        .as_secs_f64()
}
